api/indexer/indexer.go

```diff
@@ -8,230 +8,256 @@ import (
+	case blockWindow == 0:
+		return nil, errZeroBlockWindow
 	}
 	blockDB, err := pebble.New(filepath.Join(path, "block"), pebble.NewDefaultConfig(), prometheus.NewRegistry())
 	if err != nil {
 		return nil, err
 	}
 
-	blockIDCache, err := cache.NewFIFO[ids.ID, uint64](int(blockWindow))
-	if err != nil {
-		return nil, err
-	}
-	blockCache, err := cache.NewFIFO[uint64, *chain.ExecutedBlock](int(blockWindow))
-	if err != nil {
-		return nil, err
-	}
 	i := &Indexer{
 		blockDB:            blockDB,
-		blockIDToHeight:    blockIDCache,
-		blockHeightToBlock: blockCache,
+		blockIDToHeight:    make(map[ids.ID]uint64, int(blockWindow)),
+		blockHeightToBlock: make(map[uint64]*chain.ExecutedBlock, int(blockWindow)),
+		txCache:            make(map[ids.ID]cachedTransaction),
 		blockWindow:        blockWindow,
 		parser:             parser,
-		txDB:               txDB,
+		lastHeight:         math.MaxUint64,
+		cachedBlocks:       make([]uint64, 0, int(blockWindow)),
 	}
+
 	return i, i.initBlocks()
 }
 
 func (i *Indexer) initBlocks() error {
+	// Return immediately if the db is empty
+	hasLastHeight, err := i.blockDB.Has(latestBlockHeightKey)
+	if err != nil || !hasLastHeight {
+		return err
+	}
+
+	lastHeight, err := i.blockDB.Get(latestBlockHeightKey)
+	if err != nil {
+		return err
+	}
+
+	i.lastHeight = binary.BigEndian.Uint64(lastHeight)
+
+	if i.lastHeight > i.blockWindow {
+		lastRetainedHeight := i.lastHeight - i.blockWindow
+		lastRetainedHeightKey := blockEntryKey(lastRetainedHeight)
+		firstBlkKey := blockEntryKey(0)
+		if err := i.blockDB.DeleteRange(firstBlkKey, lastRetainedHeightKey); err != nil {
+			return err
+		}
+	}
+
 	// Load blockID <-> height mapping
-	iter := i.blockDB.NewIterator()
+	iter := i.blockDB.NewIteratorWithPrefix(blockEntryKeyPrefix)
 	defer iter.Release()
 
-	var lastHeight uint64
 	for iter.Next() {
 		value := iter.Value()
 		blk, err := chain.UnmarshalExecutedBlock(value, i.parser)
 		if err != nil {
 			return err
 		}
-
-		i.blockIDToHeight.Put(blk.Block.GetID(), blk.Block.Hght)
-		i.blockHeightToBlock.Put(blk.Block.Hght, blk)
-		lastHeight = blk.Block.Hght
+		i.insertBlockIntoCache(blk)
 	}
 	if err := iter.Error(); err != nil {
 		return err
 	}
 	iter.Release()
-
-	i.lastHeight.Store(lastHeight)
-
-	if lastHeight > i.blockWindow {
-		lastRetainedHeight := lastHeight - i.blockWindow
-		lastRetainedHeightBytes := binary.BigEndian.AppendUint64(nil, lastRetainedHeight)
-		if err := i.blockDB.DeleteRange(nil, lastRetainedHeightBytes); err != nil {
-			return err
-		}
-	}
-
 	return nil
 }
 
 func (i *Indexer) Notify(_ context.Context, blk *chain.ExecutedBlock) error {
-	if err := i.storeTransactions(blk); err != nil {
-		return err
-	}
-	return i.storeBlock(blk)
+	i.mu.Lock()
+	evictedBlockHeights := i.insertBlockIntoCache(blk)
+	i.mu.Unlock()
+
+	return i.storeBlock(blk, evictedBlockHeights)
```

Comment:
Ik I commented that we should return this data to make the flow more clear, but looking now, it seems we should not need this at all because we can calculate what database entry (if any) should be deleted within `i.storeBlock(...)`